Add authentication to the Preview Web UI

koszti opened the pull request:
## Description

Authentication methods for the preview UI. Partially addressing https://github.com/trinodb/trino/issues/22697

## Technical details

Introducing three new API endpoints designed for enhanced Single Page App compatibility:
* `GET: /ui/preview/auth/info`: Provides a lightweight JSON response containing the essential details for authentication type, user information, and login form rendering (if needed). Unlike the current approach, it eliminates the need for server-side HTML rendering and string manipulation.. Example:
```
$ curl -k https://127.0.0.1:8443/ui/preview/auth/info
{"authType":"form","isPasswordAllowed":true,"isAuthenticated":false}
```
* `POST: /ui/preview/auth/login`: Functionally equivalent to `POST: /ui/login`, but it exchanges data via JSON instead of HTTP forms and HTML responses.
* `GET: /ui/preview/auth/logout`: Functionally equivalent to `GET: /ui/logout`  but it exchanges data via JSON instead of HTTP redirects and HTML response

## Supported Authentication Types

* [x] `INSECURE`
* [x] `FORM` (via HTTP)
* [x] `FORM` (via HTTPS)
* [x] `FIXED`
* [x] `OAUTH2` (tested with Okta)
* [x] `JWT`: - (tested by ingesting jwt tokens into HTTP header as auth bearer token using mitmproxy)
* [x] `KERBEROS`
* [x] `CERTIFICATE`

## Screenshots

**1. `INSECURE` auth type for development. No password.**
![image](https://github.com/user-attachments/assets/71748ef2-e75c-4c33-815b-460fd42a2b86)

**2. `FORM` auth type via HTTPS**
![image](https://github.com/user-attachments/assets/fe449c6c-17c9-418a-979e-d2e1bed91338)

**3. `FIXED`**
![image](https://github.com/user-attachments/assets/c051d49e-2d68-4bfe-9df0-a21da98771fe)

**4. `OAUTH2` Okta as IdP**
![image](https://github.com/user-attachments/assets/e2ea4f60-b817-4667-999e-26d654175f23)

**NOTE:** In `oauth2`, the IdP login form is displayed (such as okta login page), whereas for `fixed` and protocol-based authenticators (`jwt`, `certifacte` and `kerberos`), users are redirected to the main page upon successful authentication without being shown a login form.

**5. Periodically fetching `/ui/api/stats` endpoint after a successful login**
![image](https://github.com/user-attachments/assets/9af85431-d5bc-4390-a55b-223e72b8d3e9)

wendigo commented:
@koszti is this ready for the review? :)

koszti commented:
> @koszti is this ready for the review? :)

yes, it's ready

wendigo commented on core/trino-main/src/main/java/io/trino/server/ui/AuthInfo.java:
```diff
@@ -0,0 +1,51 @@
+@Immutable
+public record AuthInfo(String authType, boolean isPasswordAllowed, boolean isAuthenticated, Optional<String> username)
+{
+    @Override
```
the @JsonProperty annotations are redundant.

koszti replied:
fixed

wendigo commented on core/trino-main/src/main/java/io/trino/server/ui/AuthInfo.java:
```diff
@@ -0,0 +1,51 @@
+import java.util.Optional;
+
+@Immutable
+public record AuthInfo(String authType, boolean isPasswordAllowed, boolean isAuthenticated, Optional<String> username)
```
please add a constructor validating these properties

```
public AuthInfo {
    requireNonNull(authType, "authType is null");
    requireNonNull(username, "username is null");
}
```

koszti replied:
fixed by you :)

wendigo commented on core/trino-main/src/main/java/io/trino/server/ui/AuthPreviewResource.java:
```diff
@@ -0,0 +1,23 @@
+
+public interface AuthPreviewResource
+{
+    AuthInfo getAuthInfo(ContainerRequestContext request, @Context SecurityContext securityContext);
```
I don't think that this interface is needed unless we move

```
@ResourceSecurity(WEB_UI)
    @GET
    @Path(UI_PREVIEW_AUTH_INFO)
    @Produces(APPLICATION_JSON)
```
to it

koszti replied:
deleted this interface because it doesn't like `ContainerRequestContext` being passed as parameter when jaxrs annotations moved to the interface.

wendigo commented on core/trino-main/src/main/java/io/trino/server/ui/FixedUiAuthenticatorModule.java:
```diff
@@ -18,14 +18,25 @@
 {
+    private final boolean isPreviewEnabled;
+
+    public FixedUiAuthenticatorModule(boolean isPreviewEnabled)
```
nit: uiPreviewEnabled

wendigo commented on core/trino-main/src/main/java/io/trino/server/ui/FixedUiAuthenticatorModule.java:
```diff
@@ -18,14 +18,25 @@
     public void configure(Binder binder)
     {
         binder.bind(WebUiAuthenticationFilter.class).to(FixedUserWebUiAuthenticationFilter.class).in(Scopes.SINGLETON);
         configBinder(binder).bindConfig(FixedUserWebUiConfig.class);
```
I think it would be better to move these to a PreviewUiModule that binds these conditionally:

```
class PreviewUiModule 
      extends AbstractConfigurationAwareModule
{
    String authentication = buildConfigObject(WebUiAuthenticationConfig.class).getAuthentication()

    switch (authentication) {
        // bind resources according to a type
    }
}
```

koszti replied:
can we consider it fixed by your change with the simplified binding?

wendigo commented on core/trino-main/src/main/java/io/trino/server/ui/FormWebUiAuthenticationFilter.java:
```diff
@@ -116,7 +120,8 @@ public void filter(ContainerRequestContext request)
 
         // login and logout resource is not visible to protocol authenticators
-        if ((path.equals(UI_LOGIN) && request.getMethod().equals("POST")) || path.equals(UI_LOGOUT)) {
+        if (((path.equals(UI_LOGIN) && request.getMethod().equals("POST")) || path.equals(UI_LOGOUT)) ||
```
why do we neeed that?

wendigo commented on core/trino-main/src/main/java/io/trino/server/ui/FormWebUiAuthenticationFilter.java:
```diff
@@ -227,12 +232,16 @@ public Optional<NewCookie[]> checkLoginCredentials(String username, String passw
     {
         try {
-            return MULTIPART_COOKIE.read(request.getCookies()).map(this::parseJwt);
+            Optional<String> authenticatedUsername = MULTIPART_COOKIE.read(request.getCookies()).map(this::parseJwt);
```
Why do we need that?

wendigo commented on core/trino-main/src/main/java/io/trino/server/ui/LoginPreviewResource.java:
```diff
@@ -0,0 +1,118 @@
+            throw new ForbiddenException();
+        }
+
+        return Response.ok("{}")
```
noContent()

koszti replied:
fixed

wendigo commented on core/trino-main/src/main/java/io/trino/server/ui/LoginPreviewResource.java:
```diff
@@ -0,0 +1,118 @@
+    @Produces(APPLICATION_JSON)
+    public Response logout(@Context HttpHeaders httpHeaders, @Context SecurityContext securityContext)
+    {
+        return Response.ok("{}")
```
noContent() + redirect?

koszti replied:
fixed but no redirect at the end. This endpoint is intended for APIs and the single-page frontend app so direct redirection isn’t required. The React app manages re-rendering the login form independently.

wendigo reviewed (commented):
Couple of initial comments

wendigo commented on core/trino-web-ui/src/main/resources/webapp-preview/eslint.config.js:
```diff
@@ -0,0 +1,67 @@
+/*
```
Can we extract this to a separate change?

koszti replied:
removed from this PR

wendigo commented:
@koszti I've pushed a commit to your PR that tries to simplify the bindings to reduce the scope of changes. Ideally we don't want to change existing classes at all to reduce the chance of introducing security-related bugs.

koszti commented:
> @koszti I've pushed a commit to your PR that tries to simplify the bindings to reduce the scope of changes. Ideally we don't want to change existing classes at all to reduce the chance of introducing security-related bugs.

thanks for doing that. I was struggling with that part so it was a great help. I made a few other modifications and everything is now working. I've also addressed nearly all your comments and rebased everything into a single commit.

wendigo commented:
@koszti I've pushed a bunch of changes - can you test them? I'm ok with the change as it is right now so I'm willing to merge it.

koszti commented:
@wendigo I'm fixing conflicts and testing your commits. I can see you've moved `@ResourceSecurity` annotations to class level which causes this exception:
```
Caused by: IllegalArgumentException: Trino resource is not annotated with @ResourceSecurity: public AuthInfo LoginPreviewResource.getAuthInfo(ContainerRequestContext,SecurityContext)
```

I've  checked other resources and seems like it's always annotated at the endpoints methods. Should I modify related classes accordingly so it'd look like? :
```
@Path("")
@Consumes(APPLICATION_JSON)
@Produces(APPLICATION_JSON)
public class LoginPreviewResource
{
...
    @GET
    @Path(UI_PREVIEW_AUTH_INFO)
    @ResourceSecurity(WEB_UI)
    public AuthInfo getAuthInfo(...)
...
```

wendigo commented:
@koszti according to the code it should work on the class-level as well

koszti commented:
> @koszti according to the code it should work on the class-level as well

let me look around why but it’s not working on class-level. I need a bit more time with fixing conflicts and to get things in sync with master. I’ll do it in the next couple of days

wendigo commented:
@koszti there is an invalid check in the `ResourceAccessType`. There should be a single `verifyNotTrinoResource` call there by the end of the method. Can you test with that change?

```
diff --git a/core/trino-main/src/main/java/io/trino/server/security/ResourceAccessType.java b/core/trino-main/src/main/java/io/trino/server/security/ResourceAccessType.java
index 9fe79418b35..b71a2d0b9e0 100644
--- a/core/trino-main/src/main/java/io/trino/server/security/ResourceAccessType.java
+++ b/core/trino-main/src/main/java/io/trino/server/security/ResourceAccessType.java
@@ -49,7 +49,6 @@ public class ResourceAccessType
             // check if the resource class has an access type declared for all methods
             accessType = resourceAccessTypeLoader.getAccessType(resourceInfo.getResourceClass());
             if (accessType.isPresent()) {
-                verifyNotTrinoResource(resourceInfo);
                 return accessType.get();
             }
             // in some cases there the resource is a nested class, so check the parent class
@@ -57,7 +56,6 @@ public class ResourceAccessType
             if (resourceInfo.getResourceClass().getDeclaringClass() != null) {
                 accessType = resourceAccessTypeLoader.getAccessType(resourceInfo.getResourceClass().getDeclaringClass());
                 if (accessType.isPresent()) {
-                    verifyNotTrinoResource(resourceInfo);
                     return accessType.get();
                 }
             }
```

koszti commented:
> @koszti there is an invalid check in the `ResourceAccessType`. There should be a single `verifyNotTrinoResource` call there by the end of the method. Can you test with that change?

yes, it does work after your changes, thanks! I'm going to continue the rest and will come soon

koszti commented:
@wendigo this is now ready.

wendigo commented:
@koszti there were unrelated changes in 3 files which I removed and extracted one change to a separate commit.

wendigo commented on core/trino-main/src/main/java/io/trino/server/security/ResourceAccessType.java:
```diff
@@ -49,15 +49,13 @@ public AccessType getAccessType(ResourceInfo resourceInfo)
             // check if the resource class has an access type declared for all methods
             accessType = resourceAccessTypeLoader.getAccessType(resourceInfo.getResourceClass());
             if (accessType.isPresent()) {
-                verifyNotTrinoResource(resourceInfo);
```
@dain this prohibited @ResourceSecurity at the class level for io.trino Jax-RS resources. Was it on purpose?

koszti commented:
@wendigo is this ready to merge, or is there anything I can do to assist with it?

wendigo commented:
@koszti I'll do a final pass and merge it :)

mosabua commented:
So awesome @koszti and @wendigo !!!!
